tar/lib/src/reader.dart

```diff
@@ -0,0 +1,442 @@
+// Copyright 2020
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:chunked_stream/chunked_stream.dart';
+import 'package:collection/collection.dart';
+import 'package:meta/meta.dart';
+
+import 'constants.dart';
+import 'exceptions.dart';
+import 'format.dart';
+import 'header.dart';
+import 'sparse_entry.dart';
+import 'stream.dart';
+import 'utils.dart';
+
+/// [TarReader] provides sequential access to the TAR files in a TAR archive.
+/// It is designed to read from a stream and to spit out substreams for
+/// individual file contents in order to minimize the amount of memory needed
+/// to read each archive where possible.
+@sealed
+class TarReader {
+  /// A chunked stream iterator to enable us to get our data 512 bytes at a time.
+  final ChunkedStreamIterator<int> _chunkedStream;
+
+  /// The [TarHeader] for the current file.
+  TarHeader header;
+
+  /// The current file contents.
+  Stream<List<int>> contents;
+
+  TarReader(Stream<List<int>> tarStream)
+      : _chunkedStream = ChunkedStreamIterator(tarStream) {
+    ArgumentError.checkNotNull(tarStream, 'tarStream');
+  }
+
+  /// Skip the next [_skipNext] elements when reading in the stream.
+  int _skipNext = 0;
+
+  /// Iterates through the TAR archive for the next file, returning `true` if
+  /// such a file is found, and `false` otherwise.
+  Future<bool> next() async {
+    var paxHeaders = <String, String>{};
+    var gnuLongName = '';
+    var gnuLongLink = '';
+
+    var format = TarFormat.USTAR |
+        TarFormat.PAX |
+        TarFormat.GNU |
+        TarFormat.V7 |
+        TarFormat.STAR;
+
+    /// Externally, [next] iterates through the tar archive as if it is a series
+    /// of files. Internally, the tar format often uses fake "files" to add meta
+    /// data that describes the next file. These meta data "files" should not
+    /// normally be visible to the outside. As such, this loop iterates through
+    /// one or more "header files" until it finds a "normal file".
+    while (true) {
+      if (_skipNext > 0) {
+        await _chunkedStream.read(_skipNext);
+        _skipNext = 0;
+      }
+
+      /// The discarding of the remainder of the previous file should
+      /// already be handled by [ChunkedStreamIterator].
+      var rawHeader = await _chunkedStream.read(blockSize);
+
+      header = await _readHeader(rawHeader);
```

Comment:
We might actually be updating the header in multiple steps..

I would suggest that while the future from `TarReader.next()` has not been completed, maybe calling the getter for `TarReader.header` should throw  a `StateError`. Rather than the behavior we have here where if you access `TarReader.header` while `await TarReader.next()` is still running, then you might be accessing an partially read header :see_no_evil:

Reply:
I opted to change it to a `Future` that completes right before `next` completes, in order to keep a behaviour that is somewhat like that of `StreamIterator` (i.e. `null` before the first `next` is called, and `null` when the stream is complete), while preventing users from reading it while `await TarReader.next()` is still running! 

Let me know if you think there are issues with this approach